include/pybind11/pybind11.h

```diff
@@ -2612,13 +2612,95 @@ void print(Args &&...args) {
 }
 
 error_already_set::~error_already_set() {
-    if (m_type) {
+    if (!(m_type || m_value || m_trace)) {
+        return; // Avoid gil and scope overhead if there is nothing to release.
+    }
+    gil_scoped_acquire gil;
+    error_scope scope;
+    m_type.release().dec_ref();
+    m_value.release().dec_ref();
+    m_trace.release().dec_ref();
+}
+
+error_already_set::error_already_set(const error_already_set &e) noexcept
```

Comment:
This should be noexcept, but technically the string constructor could throw an allocation error. The std lib solution to this is to reference count the what() cached str (we could do that with a shared ptr). That's very much an edge case and may not be worth the extra machinery.

Reply:
Also, I think gil_scoped_acquire can throw a pybind11_fail under certain circumstances.